src/conduits/RwaOutputConduit3.sol

```diff
@@ -0,0 +1,379 @@
+     * @notice Whitelist `who` address for `pick`.
+     * @param who The user address.
+     */
+    function kiss(address who) external auth {
+        bud[who] = 1;
+        emit Kiss(who);
+    }
+
+    /**
+     * @notice Remove `who` address from `pick` whitelist.
+     * @param who The user address.
+     */
+    function diss(address who) external auth {
+        if (to == who) to = address(0);
+        bud[who] = 0;
+        emit Diss(who);
+    }
+
+    /*//////////////////////////////////
+               Administration
+    //////////////////////////////////*/
+
+    /**
+     * @notice Updates a contract parameter.
+     * @param what The changed parameter name. `"quitTo"`
+     * @param data The new value of the parameter.
+     */
+    function file(bytes32 what, address data) external auth {
+        if (what == "quitTo") {
+            quitTo = data;
+        } else {
+            revert("RwaOutputConduit3/unrecognised-param");
+        }
+
+        emit File(what, data);
+    }
+
+    /**
+     * @notice Sets `who` address as the recipient.
+     * @param who Recipient address.
+     * @dev `who` address should have been whitelisted using `kiss`.
+     */
+    function pick(address who) external {
+        require(can[msg.sender] == 1, "RwaOutputConduit3/not-operator");
+        require(bud[who] == 1 || who == address(0), "RwaOutputConduit3/not-bud");
+        to = who;
+        emit Pick(who);
+    }
+
+    /*//////////////////////////////////
+               Operations
+    //////////////////////////////////*/
+
+    /**
+     * @notice Method to swap DAI contract balance to GEM through PSM and push it to the recipient address.
+     * @dev `msg.sender` must have been `mate`d and `to` must have been `pick`ed.
+     */
+    function push() external isMate {
+        _doPush(dai.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Swaps the specified amount of DAI into GEM through the PSM and push it to the recipient address.
+     * @dev `msg.sender` must have been `mate`d and `to` must have been `pick`ed.
+     * @param wad DAI amount.
+     */
+    function push(uint256 wad) external isMate {
+        _doPush(wad);
+    }
+
+    /**
+     * @notice Flushes out any DAI balance to `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     */
+    function quit() external isMate {
+        _doQuit(dai.balanceOf(address(this)));
+    }
+
+    /**
+     * @notice Flushes out the specified amount of DAI to the `quitTo` address.
+     * @dev `msg.sender` must have received push access through `mate()`.
+     * @param wad DAI amount.
+     */
+    function quit(uint256 wad) external isMate {
+        _doQuit(wad);
+    }
+
+    /**
+     * @notice Flushes out any GEM balance to `usr` address.
+     * @param usr Destination address
+     * @dev Can be called only by admin.
+     */
+    function yank(address usr) external auth {
+        uint256 amt = gem.balanceOf(address(this));
```

Comment:
In which case the output conduit  would have gem balance? It is very explicit that whatever amount we are getting from the PSM is sent to the recipient.
The only thing that might get stuck in normal operation (in case there are not funds in the PSM) is the DAI sent by the `urn`.

Reply:
If someone just transfers GEM to conduit, that's only the case when we can have some GEM balance

Reply:
Should we just do something more generic for admin? Excepting you see that that is a common error path that could happen.

Reply:
Maybe we can simplify this contract like this:

- Remove `quit()` and `quitTo`.
  - Most crypto brokers will generate different addresses arbitrarily for different assets throughout time, so most likely we won't have a `quitTo` when the deal is onborarded.
  - To set `quitTo`, we need a governance vote, so `quit`ting DAI would involve governance anyways.
- Make `yank` more general, by passing the address of the asset (`gem` or `dai`) as a parameter to it: `yank(address asset, address usr)`
  - We can explicitly check whether the parameter is `gem` or `dai` and revert otherwise
  - So in case any asset is stuck in the contract, we pass the ball to governance to release it to whatever destination we need to.

Reply:
Yeah this seems a great idea to me. The whole `quit` thing is what makes this contract more complex to understand. Removing that and giving `yank` access to both tokens would simplify understanding it a lot.

Reply:
> We can explicitly check whether the parameter is gem or dai and revert otherwise

Actually I think this is not needed. If someone sends other assets to this contract by mistake and governance wants to get it out, it should be able to.

Reply:
Done

Reply:
I think we should definitely keep the `quit` function. It's there to provide an escape hatch in general to allow the operator to cancel the flow for whatever reason, such as operational mistakes like sending the incorrect amount/too much funds, PSM is full, PSM fee has changed and is undesirable etc.
If we remove the quit function and only have an `auth`ed `yank`, then any mishap now requires a governance intervention. The idea to have the quit function was to minimize the need for governance actions in cases where the operator messes up. The quit function would in that case allow them to simply send back the tokens to the custody address. This custody address will be known upfront and is not expected to change - if it is, legal agreements must be amended, and both conduits must be updated anyways to whitelist a new address regardless. But in the general case this address is expected to stay the same and allow the operator to retrieve the tokens without any governance overhead. The quit function should only deal with the expected token, for the InputConduit that is gem (USDC), and for the OutputConduit that is always Dai.
While these are edge cases I think they are very likely to happen, and it would be nice to be able to deal with them without doing an executive vote everytime.

Reply:
Guys, I think before bringing any functionality back we should at least finish the discussion between the parts involved in this change. The comment was done just not much time ago and it didn't give me the time to react to it.

Honestly I don't think this is that needed, the chances that they mess with the amounts are low and simplicity in the contract is probably more worth it than saving a bit of time to governance to react in a low probable case.
I understand the PSM complicates things but I guess if something happens with it, Governance will still need to intervene somehow to decide how the loan or repayment will happen. Otherwise we should just make conduits that push/receive DAI or USDX instinctively (that was one of my suggestions to make something more resilient to future usage).
That's my personal opinion, we follow the simplicity path or we do something more robust.

Btw, last thing, could we please stop using force push here? It is the really anti pattern for following new changes in a collaborative PR.
If you are worried about the amount of cherry picks we will have at the end, then it would have been better my idea of pushing back `master` get a new clean PR and then squash and merge at the end when everybody is happy with the final version.
Anyway we can see the details on how to move all these changes to `master` later.

Reply:
> the chances that they mess with the amounts are low

How can you be so sure about this?

More often than not, people handling such transactions have more of an operational profile and are not experts in the technology.

We can still keep the generalized `yank` for unexpected assets being send to the contracts, but operational mistakes shouldn't bother Maker governance. It's just too expensive!

Think about the amount of dev time spent in writing and reviewing spells, the amount of time the guys from GovAlpha put into the executive and the time delegates spend on analyzing and voting in executives.

> Governance will still need to intervene somehow to decide how the loan or repayment will happen

That's precisely what we want to avoid.

If Dai is `draw`n from the `RwaUrn` and cannot be converted thought the PSM, the authorized party can simply call `quit` in the `OutputConduit` and send the Dai back to the `RwaUrn` (the `quitTo` address in the `OutputConduit`), so they can `wipe` the debt. No governance involvement required here.

If someones fat-fingers a `0` when sending USDC to the `InputConduit`, governance should not be required to intervene either. `InputConduit` will simply have a `quitTo` address that was previously approved by governance and they can claim it instantly.

Reply:
> don't think this is that needed, the chances that they mess with the amounts are low and simplicity in the contract is probably more worth it than saving a bit of time to governance to react in a low probable case.

@gbalabasquer it is not that amounts are wrong that will be the issue - it could be tin/tout fees on the PSM get changed for whatever reason making it better for the operator to directly send DAI.

I understand your perspective - but having seen the man hours spent to make amendments to HVB for what we thought would be edge cases in terms of the spell and code needed to implement potential amendments - and given that there is active discussion about PSM fee changes. I think @lollike's point is valid. 

When dealing with the issues with HVB, In meetings with @godsflaw and governance I got the message that we should try where ever possible to avoid having to deal with simple amendments where ever possible that require future spell/governance efforts to address simple issues, it would therefore be good to get his perspective here.

Reply:
Although I agree with @gbalabasquer (who btw is probably the most aware of what spells overhead is) that governance can fix mistakes, if you do leave `quit`, consider just leaving one verson of it per file - 

The `push` version with `amt`/`wad` was added originally because we wanted to support not processing all the existing dai/gem balance, in case some already existed there and was not related to the transaction. Although just having the version with `amt`/`wad` was sufficient, the version that processes the entire balance was left since it was expected to be used a lot. In the case of `quit`, since it is not expected to be used a lot, we can probably simplify things and leave just the version with `wad`/amt`.

Reply:
Ah another reason for having two `push` functions is that the psm might not have enough funds to swap, or we might be limited by debt ceilings, in which cases we might want a specific amount as the whole balance would revert.
Those are also not relevant for `quit`, which even more makes me think we should not have two `quit` functions.

Reply:
If we're are ok with potentially having to handle a failure scenario through an executive spell, then I'm fine with just having an authed `yank` and no `quit` function.

Reply:
> Why are we putting conduits if they are gonna send DAI directly to the urn?

This is an edge case, not the normal operational flow.
From a legal perspective, the loan never reached its destination, so it's not a repayment.

Reply:
I'm talking about the path were the user sends dozens of millions of USDC to the conduit, then "suddenly" realizes that can not use the PSM for any reason, so now they want to pull, buy the DAI themselves (probably at a higher price) and send directly to the urn? I think at that case governance will need to act anyway to give an alternative that allows them to repay.
Anyway at this point let's do whatever you prefer, I just find the way the contract is done now pretty hacky, I would tend to choose simplicity or go with more robustness and leave the user to fully operate back and forth with DAI or USDC as they want.
Closing this.